### PR TITLE
Fix using update_only_diff on nullable primitive properties

### PR DESCRIPTION
**Pull request by tgoyne**
`table.get<int64_t>()` and friends don't work on nullable int properties and would hit an assertion failure. Fix this by switching to using `switch_on_type()` which dispatches to `Optional<T>` where applicable.

### CODE REVIEW

File: src/list.hpp [hunk 1]
```diff
@@ -270,8 +270,7 @@ void List::assign(Context& ctx, T&& values, bool update, bool update_only_diff)
             index++;
         });
         while (index < sz) {
-            remove(index);
-            sz--;
+            remove(--sz);
```
Comment (ironage) [inline src/list.hpp]: nice optimisation

File: src/object_accessor.hpp [hunk 2]
```diff
@@ -69,14 +69,36 @@ ValueType Object::get_property_value(ContextType& ctx, StringData prop_name)
...
+    bool try_update;
+    bool update_only_diff;
+    bool is_default;
+
+    void operator()(RowExpr*)
+    {
+        ContextType child_ctx(ctx, property);
+        auto curr_link = row.get_link(col);
+        auto link = child_ctx.template unbox<RowExpr>(value, true, try_update, update_only_diff, curr_link);
+        if (!update_only_diff || curr_link != link.get_index()) {
```
Comment (ironage) [inline src/object_accessor.hpp]: I think null links require special handling? Or is there a test that proves they don't?

Comment (tgoyne) [reply src/object_accessor.hpp]: They're handled before we get here in both the old and new versions of the code.

File: src/object_accessor.hpp [hunk 3]
```diff
@@ -114,42 +136,9 @@ void Object::set_property_value_impl(ContextType& ctx, const Property &property,
...
-            do_update_value<Timestamp>(ctx, table, value, col, row, update_only_diff, is_default);
-            break;
-        case PropertyType::Any:
-            throw std::logic_error("not supported");
-        default:
-            REALM_COMPILER_HINT_UNREACHABLE();
-    }
+    ValueUpdater<ValueType, ContextType> updater{ctx, property, value,
+        table.get(row),col, try_update, update_only_diff, is_default};
+    switch_on_type(property.type, updater);
```
Comment (kneth) [inline src/object_accessor.hpp]: Nice simplification!

**Pull request merged**